File: src/state-dir.ts
import { mkdirSync, existsSync } from "node:fs";
import { join } from "node:path";
import { homedir, platform } from "node:os";

/**
 * Resolves the shared runtime state directory for AgentBridge.
 *
 * macOS:  ~/Library/Application Support/AgentBridge
 * Linux:  ${XDG_STATE_HOME:-~/.local/state}/agentbridge
 * Override: AGENTBRIDGE_STATE_DIR env var
 *
 * This directory stores daemon pid, lock, status, ports, and logs.
 * It is NOT for project-level config (that lives in .agentbridge/).
 */
export class StateDirResolver {
  private readonly stateDir: string;

  constructor(envOverride?: string) {
    const override = envOverride ?? process.env.AGENTBRIDGE_STATE_DIR;
    if (override) {
      this.stateDir = override;
    } else if (platform() === "darwin") {
      this.stateDir = join(homedir(), "Library", "Application Support", "AgentBridge");
    } else {
      const xdgState = process.env.XDG_STATE_HOME ?? join(homedir(), ".local", "state");
      this.stateDir = join(xdgState, "agentbridge");
    }
  }

  /** Ensure the state directory exists. */
  ensure(): void {
    if (!existsSync(this.stateDir)) {
      mkdirSync(this.stateDir, { recursive: true });
    }
  }

  get dir(): string {
    return this.stateDir;
  }

  get pidFile(): string {
    return join(this.stateDir, "daemon.pid");
  }

  get lockFile(): string {
    return join(this.stateDir, "daemon.lock");
  }

  get statusFile(): string {
    return join(this.stateDir, "status.json");
  }

  get portsFile(): string {
    return join(this.stateDir, "ports.json");
  }

  get logFile(): string {
    return join(this.stateDir, "agentbridge.log");
  }
}

File: internal/statedir/statedir.go
// Package statedir resolves the shared runtime state directory for Gossip.
package statedir

import (
	"os"
	"path/filepath"
	"runtime"
)

// StateDir wraps a resolved directory.
type StateDir struct {
	dir string
}

// New resolves the state directory, honoring GOSSIP_STATE_DIR and XDG.
func New(override string) *StateDir {
	var dir string
	if override != "" {
		dir = override
	} else if envOverride := os.Getenv("GOSSIP_STATE_DIR"); envOverride != "" {
		dir = envOverride
	} else {
		home, err := os.UserHomeDir()
		if err != nil {
			home = os.TempDir()
		}
		if runtime.GOOS == "darwin" {
			dir = filepath.Join(home, "Library", "Application Support", "Gossip")
		} else {
			xdg := os.Getenv("XDG_STATE_HOME")
			if xdg == "" {
				xdg = filepath.Join(home, ".local", "state")
			}
			dir = filepath.Join(xdg, "gossip")
		}
	}
	return &StateDir{dir: dir}
}

// Ensure creates the directory if needed.
func (s *StateDir) Ensure() error { return os.MkdirAll(s.dir, 0o755) }

// Dir returns the resolved path.
func (s *StateDir) Dir() string { return s.dir }

// PidFile returns the daemon PID file path.
func (s *StateDir) PidFile() string { return filepath.Join(s.dir, "daemon.pid") }

// LockFile returns the startup lock file path.
func (s *StateDir) LockFile() string { return filepath.Join(s.dir, "daemon.lock") }

// StatusFile returns the daemon status JSON file path.
func (s *StateDir) StatusFile() string { return filepath.Join(s.dir, "status.json") }

// LogFile returns the combined log file path.
func (s *StateDir) LogFile() string { return filepath.Join(s.dir, "gossip.log") }

// KilledFile returns the sentinel indicating intentional shutdown.
func (s *StateDir) KilledFile() string { return filepath.Join(s.dir, "killed") }

// TuiPidFile returns the managed TUI PID file path.
func (s *StateDir) TuiPidFile() string { return filepath.Join(s.dir, "codex-tui.pid") }
